File: tests/fixtures/features.py
# Test various features

from stdio import *

# Pointers and arrays
def test_pointers() -> void:
    x: int = 5
    px: -int = _.x                     # &x

    v: int = px._                      # *px
    px._ = 10                          # *px = 10

    arr: int[10]
    p0: -int = _.arr[0]                # &arr[0]

    # Pointer arithmetic
    pp: --int = _.px                   # int **pp = &px
    pp._._ = 42                        # **pp = 42

# Arrays
def test_arrays() -> void:
    arr: int[5] = [1, 2, 3, 4, 5]

    matrix: int[2][3] = [
        [1, 2, 3],
        [4, 5, 6],
    ]

    matrix[1][2] = 42

    # Designated initializer
    sparse: int[10] = {0: 1, 5: 6, 9: 10}

# Qualifiers
def test_qualifiers() -> void:
    x: const[int] = 5
    flag: volatile[int]
    u: unsigned[int] = 42
    big: unsigned[long[long]] = 1000000

# Casts
def test_casts() -> void:
    i: int = [int](3.14)
    vp: -void
    p: -int = [-int](vp)
    c: char = [char](65)

# Increment/decrement
def test_inc_dec() -> void:
    i: int = 0
    i ** _                             # i++
    _ ** i                             # ++i
    i // _                             # i--
    _ // i                             # --i

    arr: int[10]
    arr[i ** _] = 5                    # arr[i++] = 5

# Control flow
def test_control() -> void:
    x: int = 5

    # Regular if
    if x > 0:
        printf("positive\n")
    elif x < 0:
        printf("negative\n")
    else:
        printf("zero\n")

    # While
    while x < 10:
        x += 1

    # For loop
    for i in int(i := 0)(i < 5)(i ** _):
        printf("%d\n", i)

    # Do-while
    while ():
        x ** _
        if x < 20:
            continue

# Preprocessor
if [DEBUG]:
    printf("Debug mode\n")

# Macros
BUFFER_SIZE: macro = 1024
PI: macro = 3.14159

def SQUARE(x):
    x * x

def MIN(a, b):
    a if a < b else b

# Struct with designated initializer
class Point:
    x: int
    y: int

def test_struct() -> void:
    p: Point = Point(10, 20)           # {10, 20}
    q: Point(x=5, y=10)                # {.x = 5, .y = 10}
    p.x = 3

# Union
class Data(Union):
    i: int
    f: float
    c: char

def test_union() -> void:
    d: Data
    d.i = 42
    d.f = 3.14

# Enum
class Color(Enum):
    RED = 0
    GREEN = 1
    BLUE = 2

def test_enum() -> void:
    c: Color = RED
    if c == GREEN:
        printf("green\n")

# Goto and labels
def test_goto() -> void:
    i: int = 0

    LOOP: label
    if i > 10:
        raise END
    i ** _
    raise LOOP

    END: label
    printf("done\n")

File: tests/golden_outputs/features.c
#include <stdio.h>
void test_pointers(void) {
    int x = 5;
    int *px = &x;
    int v = (*px);
    (*px) = 10;
    int arr[10];
    int *p0 = &arr[0];
    int **pp = &px;
    (*(*pp)) = 42;
}
void test_arrays(void) {
    int arr[5] = {1, 2, 3, 4, 5};
    int matrix[2][3] = {{1, 2, 3}, {4, 5, 6}};
    matrix[1][2] = 42;
    int sparse[10] = {[0] = 1, [5] = 6, [9] = 10};
}
void test_qualifiers(void) {
    const int x = 5;
    volatile int flag;
    unsigned int u = 42;
    unsigned long long big = 1000000;
}
void test_casts(void) {
    int i = ((int)(3.14));
    void *vp;
    int *p = ((int*)(vp));
    char c = ((char)(65));
}
void test_inc_dec(void) {
    int i = 0;
    i++;
    ++i;
    i--;
    --i;
    int arr[10];
    arr[i++] = 5;
}
void test_control(void) {
    int x = 5;
    if (x > 0) {
        printf("positive\n");
    } else if (x < 0) {
        printf("negative\n");
    } else {
        printf("zero\n");
    }
    while (x < 10) {
        x += 1;
    }
    for (int i = 0; i < 5; i++) {
        printf("%d\n", i);
    }
    do {
        x++;
    } while (x < 20);
}
#ifdef DEBUG
printf("Debug mode\n");
#endif
#define BUFFER_SIZE 1024
#define PI 3.14159
#define SQUARE(x) ((x * x))
#define MIN(a, b) ((a < b ? a : b))
typedef struct Point {
    int x;
    int y;
} Point;
void test_struct(void) {
    Point p = {10, 20};
    struct Point q = {.x = 5, .y = 10};
    p.x = 3;
}
typedef union Data {
    int i;
    float f;
    char c;
} Data;
void test_union(void) {
    Data d;
    d.i = 42;
    d.f = 3.14;
}
typedef enum Color {
    RED = 0,
    GREEN = 1,
    BLUE = 2,
} Color;
void test_enum(void) {
    Color c = RED;
    if (c == GREEN) {
        printf("green\n");
    }
}
void test_goto(void) {
    int i = 0;
LOOP:
    if (i > 10) {
        goto END;
    }
    i++;
    goto LOOP;
END:
    printf("done\n");
}